File: src/test/resources/org/bau/benchmarks/piDigits.go
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

package main

import (
    "bufio"
    "flag"
    "fmt"
    "math/big"
    "os"
    "runtime"
    "strconv"
)

func main() {
    runtime.GOMAXPROCS(1)
    w := bufio.NewWriter(os.Stdout)
    defer w.Flush()
    line := make([]byte, 0, 10)
    var d, k int64
    var n = 10000
    flag.Parse()
    if flag.NArg() > 0 {
        n, _ = strconv.Atoi(flag.Arg(0))
    }
    for i := 1; i <= n; i++ {
        d, k = next_digit(k)
        line = append(line, byte(d) + '0')
        if len(line) == 10 {
            fmt.Fprintf(w, "%s\t:%d\n", string(line), i)
            line = line[:0]
        }
        eliminate_digit(d)
    }
}

var (
    tmp1  = big.NewInt(0)
    tmp2  = big.NewInt(0)
    y2    = big.NewInt(1)
    bigk  = big.NewInt(0)
    acc = big.NewInt(0)
    den = big.NewInt(1)
    num = big.NewInt(1)
    ten   = big.NewInt(10)
)

func next_term(k int64) int64 {
    for {
        k++
        y2.SetInt64(k * 2 + 1)
        bigk.SetInt64(k)
        tmp1.Lsh(num, 1)
        acc.Add(acc, tmp1)
        acc.Mul(acc, y2)
        den.Mul(den, y2)
        num.Mul(num, bigk)
        if acc.Cmp(num) > 0 {
            return k
        }
    }
}

func extract_digit(nth *big.Int) int64 {
    tmp1.Mul(nth, num)
    tmp2.Add(tmp1, acc)
    tmp1.Div(tmp2, den)
    return tmp1.Int64()
}

func next_digit(k int64) (int64, int64) {
    for {
        k = next_term(k)
        d3 := extract_digit(big.NewInt(3))
        d4 := extract_digit(big.NewInt(4))
        if d3 == d4 {
            return d3, k
        }
    }
}

func eliminate_digit(d int64) {
    tmp1.SetInt64(d)
    acc.Sub(acc, tmp1.Mul(den, tmp1))
    acc.Mul(acc, ten)
    num.Mul(num, ten)
}

File: src/test/resources/org/bau/benchmarks/zig/piDigits.zig
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

const std = @import("std");

var acc: std.math.big.int.Managed = undefined;
var den: std.math.big.int.Managed = undefined;
var num: std.math.big.int.Managed = undefined;
var ten: std.math.big.int.Managed = undefined;

pub fn main() !void {
    acc = try bigInit();
    defer acc.deinit();
    den = try bigInit();
    defer den.deinit();
    num = try bigInit();
    defer num.deinit();
    ten = try bigInit();
    defer ten.deinit();
    var n: usize = 10000;
    var args = std.process.args();
    _ = args.next();
    if (args.next()) |arg| {
        n = try std.fmt.parseInt(usize, arg, 10);
    }
    try acc.set(0);
    try den.set(1);
    try num.set(1);
    try ten.set(10);
    var i: usize = 0;
    var k: usize = 0;
    var stdout = std.fs.File.stdout().writerStreaming(&.{});
    while (i < n) {
        k += 1;
        try next_term(k);
        if (num.order(acc) == .gt) {
            continue;
        }
        const d3 = try extract_digit(3);
        const d4 = try extract_digit(4);
        if (d3 != d4) {
            continue;
        }
        try eliminate_digit(d3);
        try stdout.interface.print("{c}", .{'0' + @as(u8, @intCast(d3))});
        i += 1;
        if (i % 10 == 0) {
            try stdout.interface.print("\t:{d}\n", .{ i });
        }
    }
}

fn extract_digit(nth: usize) !usize {
    var tmp1 = try bigInit();
    defer tmp1.deinit();
    var tmp2 = try bigInit();
    defer tmp2.deinit();
    var tmp3 = try bigInit();
    defer tmp3.deinit();
    var nth1 = try bigInit();
    defer nth1.deinit();
    try nth1.set(nth);
    try tmp1.mul(&num, &nth1);
    try tmp1.add(&tmp1, &acc);
    try std.math.big.int.Managed.divTrunc(&tmp2, &tmp3, &tmp1, &den);
    return try tmp2.toInt(usize);
}

fn eliminate_digit(d: usize) !void {
    var tmp = try bigInit();
    defer tmp.deinit();
    var d1 = try bigInit();
    defer d1.deinit();
    try d1.set(d);
    try tmp.mul(&den, &d1);
    try acc.sub(&acc, &tmp);
    try acc.mul(&acc, &ten);
    try num.mul(&num, &ten);
}

fn next_term(k: usize) !void {
    var k2 = try bigInit();
    defer k2.deinit();
    try k2.set(k * 2 + 1);
    var tmp = try bigInit();
    defer tmp.deinit();
    try tmp.shiftLeft(&num, 1);
    try acc.add(&acc, &tmp);
    try acc.mul(&acc, &k2);
    try den.mul(&den, &k2);
    var k3 = try bigInit();
    defer k3.deinit();
    try k3.set(k);
    try num.mul(&num, &k3);
}

fn bigInit() !std.math.big.int.Managed {
    return try std.math.big.int.Managed.init(std.heap.c_allocator);
}
